render/drm_format_set.c

```diff
@@ -93,6 +93,26 @@ bool wlr_drm_format_set_add(struct wlr_drm_format_set *set, uint32_t format,
 	return true;
 }
 
+bool wlr_drm_format_set_copy(struct wlr_drm_format_set *dst,
+		const struct wlr_drm_format_set *src) {
+	dst->len = src->len;
+	dst->cap = src->cap;
+	dst->formats = calloc(dst->cap, sizeof(struct wlr_drm_format *));
+	if (dst->formats == NULL) {
+		return false;
+	}
+	for (size_t i = 0; i < src->len; i++) {
+		const struct wlr_drm_format *fmt = src->formats[i];
+		size_t fmt_size = sizeof(*fmt) + sizeof(fmt->modifiers[0]) * fmt->cap;
+		dst->formats[i] = malloc(fmt_size);
+		if (dst->formats[i] == NULL) {
+			return false;
```

Comment:
Forgot to free the `dst->formats[i]` that were allocated until this failure and also `dst->formats`.